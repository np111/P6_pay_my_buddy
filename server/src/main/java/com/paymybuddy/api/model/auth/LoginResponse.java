package com.paymybuddy.api.model.auth;

import com.paymybuddy.api.model.user.User;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Builder(builderClassName = "Builder")
@NoArgsConstructor
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Data
public class LoginResponse {
    private String token;
    private User user;
}
