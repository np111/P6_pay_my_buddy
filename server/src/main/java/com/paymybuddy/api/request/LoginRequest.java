package com.paymybuddy.api.request;

import com.paymybuddy.api.validation.constraint.IsEmail;
import com.paymybuddy.api.validation.constraint.IsPassword;
import javax.validation.constraints.NotNull;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

@Builder(builderClassName = "Builder")
@NoArgsConstructor
@AllArgsConstructor(access = AccessLevel.PRIVATE)
@Data
public class LoginRequest {
    @NotNull
    @IsEmail
    private String email;

    @NotNull
    @IsPassword
    private String password;
}
