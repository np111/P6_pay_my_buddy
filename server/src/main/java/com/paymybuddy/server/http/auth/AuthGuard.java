package com.paymybuddy.server.http.auth;

import com.paymybuddy.api.model.User;

public interface AuthGuard {
    boolean isAuthenticated();

    long getUserId();

    User getUser();
}
