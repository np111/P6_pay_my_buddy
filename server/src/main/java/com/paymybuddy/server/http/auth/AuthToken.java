package com.paymybuddy.server.http.auth;

import com.google.common.base.Preconditions;
import java.util.Collections;
import org.springframework.security.authentication.AbstractAuthenticationToken;

public final class AuthToken extends AbstractAuthenticationToken {
    public static AuthToken unauthenticated(String token) {
        return new AuthToken(null, token);
    }

    public static AuthToken authenticated(AuthGuard auth, String token) {
        Preconditions.checkNotNull(auth, "auth cannot be null");
        Preconditions.checkState(auth.isAuthenticated());
        return new AuthToken(auth, token);
    }

    private final AuthGuard auth;
    private final String token;

    private AuthToken(AuthGuard auth, String token) {
        super(auth == null ? null : Collections.emptyList()); // (no authorities)
        this.auth = auth;
        this.token = token;
        setAuthenticated(auth != null);
    }

    @Override
    public AuthGuard getPrincipal() {
        return auth;
    }

    @Override
    public String getCredentials() {
        return token;
    }
}
