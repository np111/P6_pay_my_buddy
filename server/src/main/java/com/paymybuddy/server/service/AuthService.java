package com.paymybuddy.server.service;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.paymybuddy.api.model.Currency;
import com.paymybuddy.api.model.user.User;
import com.paymybuddy.server.http.auth.AuthToken;
import com.paymybuddy.server.jpa.entity.UserEntity;
import com.paymybuddy.server.jpa.mapper.UserMapper;
import com.paymybuddy.server.jpa.repository.UserRepository;
import com.paymybuddy.server.util.exception.FastRuntimeException;
import java.time.ZonedDateTime;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import lombok.Data;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.RequiredArgsConstructor;
import lombok.ToString;
import org.apache.commons.validator.GenericValidator;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.data.redis.connection.lettuce.LettuceConnectionFactory;
import org.springframework.data.redis.core.HashOperations;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.serializer.Jackson2JsonRedisSerializer;
import org.springframework.data.redis.serializer.SerializationException;
import org.springframework.data.redis.serializer.StringRedisSerializer;
import org.springframework.lang.Nullable;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.authentication.CredentialsExpiredException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Service
@Scope("singleton")
public class AuthService implements InitializingBean {
    private static final String AUTH_SESS_BY_ID_HASH_KEY = "authsess.";
    private static final int AUTH_TOKEN_MAX_LEN = Long.toString(Long.MAX_VALUE).length() + 1 + new UUID(0, 0).toString().length();

    private final @Getter PasswordEncoder passwordEncoder;
    private String userNotFoundEncodedPassword;

    private final @Getter UserService userService;
    private final UserRepository userRepository;
    private final UserMapper userMapper;

    private final LettuceConnectionFactory lettuceConFactory;
    private RedisTemplate<String, ?> authSessTemplate;
    private HashOperations<String, String, AuthGuard> authSessHashOps;

    @Override
    public void afterPropertiesSet() throws Exception {
        userNotFoundEncodedPassword = passwordEncoder.encode(UUID.randomUUID().toString());

        authSessTemplate = new RedisTemplate<>();
        authSessTemplate.setConnectionFactory(lettuceConFactory);
        authSessTemplate.setKeySerializer(new StringRedisSerializer());
        authSessTemplate.setHashKeySerializer(new StringRedisSerializer());
        authSessTemplate.setHashValueSerializer(new AuthSerializer());
        authSessTemplate.afterPropertiesSet();
        authSessHashOps = authSessTemplate.opsForHash();
    }

    @Transactional
    public AuthToken login(String email, String password) {
        User user = userService.getUserByEmail(email);
        if (user == null) {
            // mitigate timing attack
            passwordEncoder.matches(password, userNotFoundEncodedPassword);
            throw new BadCredentialsException("Bad credentials");
        }
        if (!passwordEncoder.matches(password, user.getEncodedPassword())) {
            throw new BadCredentialsException("Bad credentials");
        }
        if (passwordEncoder.upgradeEncoding(user.getEncodedPassword())) {
            // upgrade password encoding
            userRepository.updatePassword(user.getId(), passwordEncoder.encode(password));
        }
        return createAuthToken(user);
    }

    private AuthToken createAuthToken(User user) {
        AuthGuard auth = new AuthGuard(user);
        auth.authService = this;
        String sessionId = UUID.randomUUID().toString();
        String token = user.getId() + "." + sessionId;
        String redisKey = AUTH_SESS_BY_ID_HASH_KEY + user.getId();
        authSessHashOps.put(redisKey, sessionId, auth);
        authSessTemplate.expire(redisKey, 30, TimeUnit.DAYS);
        return AuthToken.authenticated(auth, token);
    }

    @Nullable
    public AuthToken authenticateAuthToken(AuthToken token) {
        if (token.getCredentials().length() <= AUTH_TOKEN_MAX_LEN) {
            String[] tokenParts = token.getCredentials().split("\\.", 2);
            if (tokenParts.length == 2) {
                String userId = tokenParts[0];
                String sessionId = tokenParts[1];
                AuthGuard auth = authSessHashOps.get(AUTH_SESS_BY_ID_HASH_KEY + userId, sessionId);
                if (auth != null) {
                    return AuthToken.authenticated(auth, token.getCredentials());
                }
            }
        }
        throw new CredentialsExpiredException("Invalid or expired auth token");
    }

    public void destroyAuthToken(AuthToken token) {
        if (token.getCredentials().length() <= AUTH_TOKEN_MAX_LEN) {
            String[] tokenParts = token.getCredentials().split("\\.", 2);
            if (tokenParts.length == 2) {
                String userId = tokenParts[0];
                String sessionId = tokenParts[1];
                authSessHashOps.delete(AUTH_SESS_BY_ID_HASH_KEY + userId, sessionId);
            }
        }
    }

    @Transactional
    public User register(String name, String email, String password, Currency defaultCurrency) throws IllegalNameException, IllegalEmailException, EmailAlreadyRegisteredException {
        name = validateAndNormalizeNewName(name);
        email = validateAndNormalizeNewEmail(email);
        validateNewPassword(password);

        UserEntity userEntity = new UserEntity();
        userEntity.setName(name);
        userEntity.setEmail(email);
        userEntity.setEncodedPassword(passwordEncoder.encode(password));
        userEntity.setDefaultCurrency(defaultCurrency);
        try {
            userRepository.save(userEntity);
        } catch (DataIntegrityViolationException ex) {
            throw new EmailAlreadyRegisteredException();
        }
        return userMapper.toUser(userEntity);
    }

    private String validateAndNormalizeNewName(String name) {
        name = userService.normalizeName(name);
        if (name == null) {
            throw new IllegalNameException();
        }
        // TODO: Advanced validation
        return userService.normalizeName(name);
    }

    private String validateAndNormalizeNewEmail(String email) {
        email = userService.normalizeEmail(email);
        if (email == null || !GenericValidator.isEmail(email)) {
            throw new IllegalEmailException();
        }
        return email;
    }

    private void validateNewPassword(String password) {
        // TODO(high): validate password preconditions (minimum length, symbols?, etc)
    }

    @NoArgsConstructor
    @Data
    @ToString(of = {"userId"})
    static class AuthGuard implements com.paymybuddy.server.http.auth.AuthGuard {
        private long userId;
        private ZonedDateTime loginDate;

        private transient AuthService authService;
        private transient User user;

        public AuthGuard(User user) {
            this.userId = user.getId();
            this.user = user;
            this.loginDate = ZonedDateTime.now();
        }

        @Override
        public boolean isAuthenticated() {
            return userId != 0L;
        }

        public User getUser() {
            if (user == null && isAuthenticated()) {
                user = authService.getUserService().getUserById(userId);
                if (user == null) {
                    throw new RuntimeException("user not found (id: " + userId + ")");
                }
            }
            return user;
        }
    }

    class AuthSerializer extends Jackson2JsonRedisSerializer<AuthGuard> {
        public AuthSerializer() {
            super(AuthGuard.class);
            ObjectMapper mapper = new ObjectMapper();
            mapper.setVisibility(PropertyAccessor.ALL, JsonAutoDetect.Visibility.NONE);
            mapper.setVisibility(PropertyAccessor.FIELD, JsonAutoDetect.Visibility.ANY);
            mapper.configure(MapperFeature.PROPAGATE_TRANSIENT_MARKER, true);
            mapper.configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, true);
            mapper.configure(SerializationFeature.WRITE_DATE_TIMESTAMPS_AS_NANOSECONDS, false);
            mapper.registerModule(new JavaTimeModule());
            setObjectMapper(mapper);
        }

        @Override
        public AuthGuard deserialize(byte[] data) throws SerializationException {
            AuthGuard auth = super.deserialize(data);
            if (auth != null) {
                auth.authService = AuthService.this;
            }
            return auth;
        }
    }

    public static class IllegalNameException extends FastRuntimeException {
    }

    public static class IllegalEmailException extends FastRuntimeException {
    }

    public static class EmailAlreadyRegisteredException extends FastRuntimeException {
    }
}
