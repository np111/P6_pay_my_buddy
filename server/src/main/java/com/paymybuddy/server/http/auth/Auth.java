package com.paymybuddy.server.http.auth;

import com.paymybuddy.api.model.User;
import com.paymybuddy.server.service.AuthService;
import lombok.AccessLevel;
import lombok.Data;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;

@NoArgsConstructor
@Data
@ToString(of = {"userId"})
public class Auth {
    private long userId;

    private transient @Getter(AccessLevel.NONE) AuthService authService;
    private transient User user;

    public Auth(long userId) {
        this.userId = userId;
    }

    public Auth(User user) {
        this(user.getId());
        this.user = user;
    }

    public void setAuthService(AuthService authService) {
        this.authService = authService;
    }

    public User getUser() {
        if (user == null) {
            return user = authService.getUserService().getUserById(userId);
        }
        return user;
    }
}
