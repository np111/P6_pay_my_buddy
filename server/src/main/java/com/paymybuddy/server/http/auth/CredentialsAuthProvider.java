package com.paymybuddy.server.http.auth;

import com.paymybuddy.server.service.AuthService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Component
@Scope("singleton")
public class CredentialsAuthProvider implements AuthenticationProvider {
    private final AuthService authService;

    @Override
    public AuthToken authenticate(Authentication authentication) throws AuthenticationException {
        UsernamePasswordAuthenticationToken credentials = (UsernamePasswordAuthenticationToken) authentication;
        return authService.login(determineEmail(credentials), determinePassword(credentials));
    }

    private String determineEmail(Authentication authentication) {
        return authentication.getPrincipal() == null ? "NONE_PROVIDED" : authentication.getName();
    }

    private String determinePassword(Authentication authentication) {
        return authentication.getCredentials().toString();
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return UsernamePasswordAuthenticationToken.class.isAssignableFrom(authentication);
    }
}
