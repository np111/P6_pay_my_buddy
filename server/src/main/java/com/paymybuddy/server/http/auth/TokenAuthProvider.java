package com.paymybuddy.server.http.auth;

import com.paymybuddy.server.service.AuthService;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Scope;
import org.springframework.security.authentication.AuthenticationProvider;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.stereotype.Component;

@RequiredArgsConstructor(onConstructor = @__(@Autowired))
@Component
@Scope("singleton")
public class TokenAuthProvider implements AuthenticationProvider {
    private final AuthService authService;

    @Override
    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
        return authService.authenticateAuthToken((AuthToken) authentication);
    }

    @Override
    public boolean supports(Class<?> authentication) {
        return AuthToken.class.isAssignableFrom(authentication);
    }
}
