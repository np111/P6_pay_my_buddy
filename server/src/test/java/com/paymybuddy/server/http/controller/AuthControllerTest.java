package com.paymybuddy.server.http.controller;

import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.paymybuddy.api.request.auth.LoginRequest;
import com.paymybuddy.auth.AuthService;
import com.paymybuddy.auth.AuthToken;
import com.paymybuddy.auth.provider.CredentialsAuthProvider;
import com.paymybuddy.auth.provider.TokenAuthProvider;
import com.paymybuddy.server.mock.MockAuthGuard;
import com.paymybuddy.server.mock.TestControllerConfig;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.context.annotation.Import;
import org.springframework.http.MediaType;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.BadCredentialsException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.test.web.servlet.MockMvc;

import static com.paymybuddy.server.mock.MockMvcSnapshot.toMatchSnapshot;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;

@WebMvcTest(AuthController.class)
@Import(TestControllerConfig.class)
@ExtendWith(SnapshotExtension.class)
class AuthControllerTest {
    @MockBean
    private AuthService authService;
    @MockBean
    private AuthenticationManager authenticationManager;
    @MockBean
    private CredentialsAuthProvider credentialsAuthProvider;
    @MockBean
    private TokenAuthProvider tokenAuthProvider;

    @Autowired
    private MockMvc mockMvc;
    @Autowired
    private ObjectMapper objectMapper;

    @Test
    void login() throws Exception {
        when(authService.login(any(), any())).thenAnswer(m -> {
            String email = m.getArgument(0);
            String password = m.getArgument(1);
            if ("email@domain.tld".equals(email) && "password".equals(password)) {
                return AuthToken.authenticated(MockAuthGuard.create(), "SECRET");
            }
            throw new BadCredentialsException("Bad credentials");
        });
        toMatchSnapshot(
                mockMvc.perform(post("/auth/login")
                        .header("x-auth-token", "anonymous")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(LoginRequest.builder()
                                .email("email@domain.tld")
                                .password("password")
                                .build())))
                        .andReturn(),
                mockMvc.perform(post("/auth/login")
                        .header("x-auth-token", "anonymous")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(LoginRequest.builder()
                                .email("bad_email@domain.tld")
                                .password("password")
                                .build())))
                        .andReturn(),
                mockMvc.perform(post("/auth/login")
                        .header("x-auth-token", "anonymous")
                        .contentType(MediaType.APPLICATION_JSON)
                        .content(objectMapper.writeValueAsString(LoginRequest.builder()
                                .email("email@domain.tld")
                                .password("bad password")
                                .build())))
                        .andReturn()
        );
    }

    @Test
    void rememberNotLogged() throws Exception {
        when(authenticationManager.authenticate(any())).thenThrow(new BadCredentialsException("Bad credentials"));
        toMatchSnapshot(
                mockMvc.perform(get("/auth/remember")
                        .header("x-auth-token", "expired-token"))
                        .andReturn()
        );
    }

    @Test
    @MockAuthGuard.WithAuthToken
    void rememberLogged() throws Exception {
        toMatchSnapshot(
                mockMvc.perform(get("/auth/remember")).andReturn()
        );
    }

    @Test
    @MockAuthGuard.WithAuthToken
    void logout() throws Exception {
        String token = ((AuthToken) SecurityContextHolder.getContext().getAuthentication()).getCredentials();
        toMatchSnapshot(
                mockMvc.perform(post("/auth/logout")).andReturn()
        );
        verify(authService, times(1)).destroyAuthToken(token);
    }
}