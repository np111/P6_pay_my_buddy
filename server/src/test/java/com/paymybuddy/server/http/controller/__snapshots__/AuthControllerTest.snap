com.paymybuddy.server.http.controller.AuthControllerTest.login=[
-
> POST /auth/login
> Content-Type: application/json;charset=UTF-8
> x-auth-token: anonymous
> Content-Length: 50
{"email":"email@domain.tld","password":"password"}
< HTTP 200
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
{"token":"SECRET","user":{"id":1,"email":"1@domain.tld","name":"#1","defaultCurrency":"USD"}}
-
> POST /auth/login
> Content-Type: application/json;charset=UTF-8
> x-auth-token: anonymous
> Content-Length: 54
{"email":"bad_email@domain.tld","password":"password"}
< HTTP 403
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
{"type":"SERVICE","status":403,"code":"INVALID_CREDENTIALS","message":"Authentication failed: Bad credentials","metadata":{}}
-
> POST /auth/login
> Content-Type: application/json;charset=UTF-8
> x-auth-token: anonymous
> Content-Length: 54
{"email":"email@domain.tld","password":"bad password"}
< HTTP 403
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
{"type":"SERVICE","status":403,"code":"INVALID_CREDENTIALS","message":"Authentication failed: Bad credentials","metadata":{}}
]


com.paymybuddy.server.http.controller.AuthControllerTest.logout=[
-
> POST /auth/logout
< HTTP 204
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY

]


com.paymybuddy.server.http.controller.AuthControllerTest.rememberLogged=[
-
> GET /auth/remember
< HTTP 200
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
{"user":{"id":1,"email":"1@domain.tld","name":"#1","defaultCurrency":"USD"}}
]


com.paymybuddy.server.http.controller.AuthControllerTest.rememberNotLogged=[
-
> GET /auth/remember
> x-auth-token: expired-token
< HTTP 403
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: application/json; charset=utf-8
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< Cache-Control: no-cache, no-store, max-age=0, must-revalidate
< Pragma: no-cache
< Expires: 0
< X-Frame-Options: DENY
{"type":"SERVICE","status":403,"code":"INVALID_CREDENTIALS","message":"Authentication failed: Bad credentials","metadata":{}}
]